src/apps/eos/actions/__init__.py

```diff
@@ -0,0 +1,107 @@
+    elif action.undelegate is not None:
+        await layout.confirm_action_undelegate(ctx, action.undelegate)
+        updaters.hashupdate_action_undelegate(sha, action.undelegate)
+    elif action.refund is not None:
+        await layout.confirm_action_refund(ctx, action.refund)
+        updaters.hashupdate_action_refund(sha, action.refund)
+    elif action.transfer is not None:
+        await layout.confirm_action_transfer(ctx, action.transfer)
+        updaters.hashupdate_action_transfer(sha, action.transfer)
+    elif action.vote_producer is not None:
+        await layout.confirm_action_voteproducer(ctx, action.vote_producer)
+        updaters.hashupdate_action_voteproducer(sha, action.vote_producer)
+    elif action.update_auth is not None:
+        await layout.confirm_action_updateauth(ctx, action.update_auth)
+        updaters.hashupdate_action_updateauth(sha, action.update_auth)
+    elif action.delete_auth is not None:
+        await layout.confirm_action_deleteauth(ctx, action.delete_auth)
+        updaters.hashupdate_action_deleteauth(sha, action.delete_auth)
+    elif action.link_auth is not None:
+        await layout.confirm_action_linkauth(ctx, action.link_auth)
+        updaters.hashupdate_action_linkauth(sha, action.link_auth)
+    elif action.unlink_auth is not None:
+        await layout.confirm_action_unlinkauth(ctx, action.unlink_auth)
+        updaters.hashupdate_action_unlinkauth(sha, action.unlink_auth)
+    elif action.new_account is not None:
+        await layout.confirm_action_newaccount(ctx, action.new_account)
+        updaters.hashupdate_action_newaccount(sha, action.new_account)
+    elif action.unknown is not None:
+        await process_unknown_action(ctx, sha, action)
+    else:
+        raise ValueError("Unknown action")
+
+async def process_unknown_action(ctx, sha, action):
+    checksum = HashWriter(sha256)
+    checksum.extend(helpers.pack_variant32(action.unknown.data_size))
+    checksum.extend(action.unknown.data_chunk)
+
+    updaters.hashupdate_variant32(sha, action.unknown.data_size)
+    updaters.hashupdate_bytes(sha, action.unknown.data_chunk)
+    bytes_left = action.unknown.data_size - len(action.unknown.data_chunk)
+
+    while bytes_left != 0:
+        action = await ctx.call(EosTxActionRequest(data_size=bytes_left), *consts.action_wire_types)
+
+        if action.unknown is None:
+            raise ValueError("Bad response. Unknown struct expected.")
+
+        checksum.extend(action.unknown.data_chunk)
+        updaters.hashupdate_bytes(sha, action.unknown.data_chunk)
+
+        bytes_left -= len(action.unknown.data_chunk)
+        if bytes_left < 0:
+            raise ValueError("Bad response. Buffer overflow.")
+
+    await layout.confirm_action_unknown(ctx, action.common, checksum.get_digest())
+
+
+def check_action(action):
+    account = action.common.account
+    name = action.common.name
+
+    if account == helpers.eos_name_string_to_number('eosio'):
```

Comment:
might be a good idea to precompute these

Reply:
done

Reply:
@matejcik Hi. It is already precomputed and reworked a bit.